bigtable-client-core-parent/bigtable-hbase/src/main/java/com/google/cloud/bigtable/hbase/BigtableOAuth2Credentials.java

```diff
@@ -0,0 +1,106 @@
+ * the {@link com.google.auth.Credentials} class as it is shaded by Cloud Bigtable client.
+ *
+ * <p>Hence, customers should implement this class, which will be used for authentication. The
+ * authentication should be based on OAuth2 and must work by just including request metadata with
+ * each request at transport layer.
+ */
+public abstract class BigtableOAuth2Credentials {
+
+  /**
+   * All subclasses must implement this constructor and populate the @configuration. The
+   * configuration should be used to initialize the credentials.
+   *
+   * @param configuration The HBase configuration
+   */
+  public BigtableOAuth2Credentials(Configuration configuration) {
+    this.configuration = configuration;
+  }
+
+  /**
+   * Get the current request metadata.
+   *
+   * <p>This should be called by the transport layer on each request, and the data should be
+   * populated in headers or other context.
+   *
+   * <p>The convention for handling binary data is for the key in the returned map to end with
+   * {@code "-bin"} and for the corresponding values to be base64 encoded.
+   *
+   * <p>This class should handle caching and refreshing of the metadata associated with the request.
+   * Ideally, caching and refreshing of credentials should happen in an asynchronous non-blocking
+   * way.
+   *
+   * @param uri URI of the entry point for the request.
+   */
+  public abstract Map<String, List<String>> getRequestMetadata(URI uri) throws IOException;
+
+  /** Returns the HBase configuration used to create this object. */
+  public Configuration getConfiguration() {
+    return configuration;
+  }
+
+  /**
+   * Creates a new instance of a child of @{@link BigtableOAuth2Credentials}.
+   *
+   * @param bigtableAuthClass the child class to be instantiated
+   * @param conf HBase configuration required to configure the @bigtableAuthClass
+   * @return a new instance of @bigtableAuthClass
+   */
+  public static Credentials newInstance(
+      Class<? extends BigtableOAuth2Credentials> bigtableAuthClass, Configuration conf) {
+
+    Constructor<?> constructor = null;
+    try {
+      constructor = bigtableAuthClass.getConstructor(Configuration.class);
+    } catch (NoSuchMethodException e) {
+      throw new IllegalArgumentException(
+          "Custom credentials class ["
+              + bigtableAuthClass
+              + "] must implement a constructor with single argument of type "
+              + Configuration.class.getName()
+              + ".",
+          e);
+    }
+
+    try {
+      return new BigtableCredentialsWrapper(
+          (BigtableOAuth2Credentials) constructor.newInstance(conf));
+    } catch (Exception e) {
+      throw new IllegalArgumentException(
+          "Failed to create object of custom Credentials class ["
+              + bigtableAuthClass.getName()
+              + "].",
+          e);
+    }
+  }
+
+  protected Configuration configuration;
```

Comment:
Its weird that you have both a public getter and protected member variable for Configuration. I think it would be a lot simpler if you drop both and let endusers retain the ref from the constructor

Reply:
ok.